File: examples/python/facebook.py
"""
Cognima API - Facebook Download Examples (Python)

This file demonstrates how to use the Facebook download endpoints
to download videos with multiple quality options.
"""

import requests
from typing import List, Dict, Optional

# API Configuration
BASE_URL = 'https://cog.api.br/api/v1'
API_KEY = 'YOUR_API_KEY_HERE'

# Configure headers
headers = {
    'apikey': API_KEY
}

def download_facebook_video(url: str) -> List[Dict]:
    """
    Download Facebook video (all quality options)
    
    Args:
        url: Facebook video URL
    
    Returns:
        List of video quality options
    """
    print('\n=== Downloading Facebook Video (All Qualities) ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        videos = data['videos']
        print('✅ Download URLs Ready!\n')
        print(f"Found {len(videos)} quality options:\n")
        
        for i, video in enumerate(videos, 1):
            print(f"{i}. Resolution: {video['resolution']}")
            print(f"   URL: {video['url']}")
            print(f"   Thumbnail: {video['thumbnail']}")
            print(f"   Should Render: {'Yes' if video['shouldRender'] else 'No'}")
            print()
        
        return videos
    
    return []

def download_facebook_video_hd(url: str) -> Dict:
    """
    Download Facebook video in best quality (HD preferred)
    
    Args:
        url: Facebook video URL
    
    Returns:
        Best quality video information
    """
    print('\n=== Downloading Facebook Video (Best Quality) ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download-hd', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        video = data['video']
        all_qualities = data['allQualities']
        
        print('✅ Best Quality Download Ready!\n')
        print('🎬 Selected Quality:')
        print(f"   Resolution: {video['resolution']}")
        print(f"   URL: {video['url']}")
        print(f"   Thumbnail: {video['thumbnail']}")
        print(f"   Should Render: {'Yes' if video['shouldRender'] else 'No'}")
        print(f"\n📊 Available Qualities: {len(all_qualities)}")
        
        for i, q in enumerate(all_qualities, 1):
            print(f"   {i}. {q['resolution']}")
        
        return video
    
    return {}

def download_multiple_videos(urls: List[str]) -> List[Dict]:
    """
    Download multiple Facebook videos
    
    Args:
        urls: List of Facebook video URLs
    
    Returns:
        List of download results
    """
    print('\n=== Downloading Multiple Facebook Videos ===\n')
    
    results = []
    
    for url in urls:
        try:
            params = {'url': url}
            response = requests.get(f'{BASE_URL}/facebook/download-hd', params=params, headers=headers)
            response.raise_for_status()
            
            data = response.json()
            
            results.append({
                'url': url,
                'success': True,
                'video': data['video']
            })
            
            print(f"✅ Downloaded: {data['video']['resolution']}")
        except Exception as e:
            results.append({
                'url': url,
                'success': False,
                'error': str(e)
            })
            
            print(f"❌ Failed: {url}")
    
    successful = len([r for r in results if r['success']])
    print(f"\n📋 Downloaded {successful}/{len(urls)} videos")
    
    return results

def download_specific_quality(url: str, preferred_resolution: str) -> Dict:
    """
    Get specific quality from Facebook video
    
    Args:
        url: Facebook video URL
        preferred_resolution: Preferred resolution (e.g., "720p", "1080p")
    
    Returns:
        Video information for preferred quality
    """
    print(f'\n=== Downloading Facebook Video ({preferred_resolution}) ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        videos = data['videos']
        
        # Find preferred quality
        video = next(
            (v for v in videos if preferred_resolution.lower() in v['resolution'].lower()),
            videos[0]
        )
        
        print('✅ Download URL Ready!\n')
        print(f"Resolution: {video['resolution']}")
        print(f"URL: {video['url']}")
        print(f"Thumbnail: {video['thumbnail']}")
        
        return video
    
    return {}

def get_video_info(url: str) -> Dict:
    """
    Get video information without downloading
    
    Args:
        url: Facebook video URL
    
    Returns:
        Dictionary with video information
    """
    print('\n=== Getting Facebook Video Info ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        videos = data['videos']
        
        print('📊 Video Information:\n')
        print(f"Available Qualities: {len(videos)}")
        
        for i, video in enumerate(videos, 1):
            print(f"\n{i}. {video['resolution']}")
            print(f"   Has Thumbnail: {'Yes' if video['thumbnail'] else 'No'}")
            print(f"   Needs Rendering: {'Yes' if video['shouldRender'] else 'No'}")
        
        # Find best quality
        best_quality = (
            next((v for v in videos if '1080p' in v['resolution']), None) or
            next((v for v in videos if '720p' in v['resolution']), None) or
            videos[0]
        )
        
        print(f"\n🏆 Best Quality: {best_quality['resolution']}")
        
        return {
            'totalQualities': len(videos),
            'qualities': [v['resolution'] for v in videos],
            'bestQuality': best_quality['resolution']
        }
    
    return {}

def compare_qualities(url: str) -> List[Dict]:
    """
    Download and compare qualities
    
    Args:
        url: Facebook video URL
    
    Returns:
        List of quality comparisons
    """
    print('\n=== Comparing Video Qualities ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        videos = data['videos']
        
        print('📊 Quality Comparison:\n')
        
        comparison = []
        for video in videos:
            comp = {
                'resolution': video['resolution'],
                'hasThumbnail': bool(video['thumbnail']),
                'needsRender': video['shouldRender'],
                'isDirectLink': not video['shouldRender']
            }
            comparison.append(comp)
            
            print(f"Resolution: {comp['resolution']}")
            print(f"  Thumbnail: {'Yes' if comp['hasThumbnail'] else 'No'}")
            print(f"  Needs Render: {'Yes' if comp['needsRender'] else 'No'}")
            print(f"  Direct Link: {'Yes' if comp['isDirectLink'] else 'No'}")
            print()
        
        return comparison
    
    return []

def download_best_from_multiple_sources(urls: List[str]) -> List[Dict]:
    """
    Download best quality from multiple Facebook video URLs
    
    Args:
        urls: List of Facebook video URLs
    
    Returns:
        List of best quality videos
    """
    print('\n=== Downloading Best Quality from Multiple Sources ===\n')
    
    best_videos = []
    
    for i, url in enumerate(urls, 1):
        print(f'Processing {i}/{len(urls)}...')
        
        try:
            video = download_facebook_video_hd(url)
            best_videos.append({
                'url': url,
                'video': video,
                'success': True
            })
        except Exception as e:
            print(f'❌ Error: {str(e)}')
            best_videos.append({
                'url': url,
                'error': str(e),
                'success': False
            })
        
        print()
    
    successful = len([v for v in best_videos if v['success']])
    print(f'✅ Successfully downloaded {successful}/{len(urls)} videos')
    
    return best_videos

def filter_by_resolution(url: str, min_resolution: str = '720p') -> List[Dict]:
    """
    Filter video qualities by minimum resolution
    
    Args:
        url: Facebook video URL
        min_resolution: Minimum resolution required
    
    Returns:
        List of filtered video qualities
    """
    print(f'\n=== Filtering by Minimum Resolution ({min_resolution}) ===\n')
    print(f'URL: {url}\n')
    
    params = {'url': url}
    
    response = requests.get(f'{BASE_URL}/facebook/download', params=params, headers=headers)
    response.raise_for_status()
    
    data = response.json()
    
    if data['success']:
        videos = data['videos']
        
        # Simple resolution filtering
        resolution_order = {'1080p': 3, '720p': 2, '480p': 1, '360p': 0}
        min_value = resolution_order.get(min_resolution, 1)
        
        filtered = [
            v for v in videos
            if any(res in v['resolution'] for res in resolution_order.keys()
                   if resolution_order[res] >= min_value)
        ]
        
        print(f'Found {len(filtered)} qualities >= {min_resolution}:\n')
        for video in filtered:
            print(f"  - {video['resolution']}")
        
        return filtered
    
    return []

# ===================
# EXAMPLES
# ===================

def main():
    try:
        test_url = 'https://www.facebook.com/share/r/14RStacRcih/'
        
        # Example 1: Download all quality options
        download_facebook_video(test_url)
        
        # Example 2: Download best quality (HD preferred)
        download_facebook_video_hd(test_url)
        
        # Example 3: Download specific quality
        download_specific_quality(test_url, '720p')
        
        # Example 4: Get video information
        get_video_info(test_url)
        
        # Example 5: Compare qualities
        compare_qualities(test_url)
        
        # Example 6: Download multiple videos
        urls = [
            'https://www.facebook.com/share/r/14RStacRcih/',
            'https://fb.watch/example123/'
        ]
        download_multiple_videos(urls)
        
        # Example 7: Download best from multiple sources
        download_best_from_multiple_sources(urls)
        
        # Example 8: Filter by resolution
        filter_by_resolution(test_url, '720p')
        
    except Exception as e:
        print(f'Error: {str(e)}')
        exit(1)

if __name__ == '__main__':
    main()

File: examples/nodejs/facebook.js
/**
 * Cognima API - Facebook Download Examples (Node.js)
 * 
 * This file demonstrates how to use the Facebook download endpoints
 * to download videos with multiple quality options.
 */

const axios = require('axios');

// API Configuration
const BASE_URL = 'https://cog.api.br/api/v1';
const API_KEY = 'YOUR_API_KEY_HERE';

// Configure axios with API key
const api = axios.create({
  baseURL: BASE_URL,
  headers: {
    'apikey': API_KEY
  }
});

/**
 * Download Facebook video (all quality options)
 * @param {string} url - Facebook video URL
 */
async function downloadFacebookVideo(url) {
  try {
    console.log('\n=== Downloading Facebook Video (All Qualities) ===\n');
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/facebook/download', {
      params: { url }
    });
    
    const { success, videos } = response.data;
    
    if (success) {
      console.log('✅ Download URLs Ready!\n');
      console.log(`Found ${videos.length} quality options:\n`);
      
      videos.forEach((video, index) => {
        console.log(`${index + 1}. Resolution: ${video.resolution}`);
        console.log(`   URL: ${video.url}`);
        console.log(`   Thumbnail: ${video.thumbnail}`);
        console.log(`   Should Render: ${video.shouldRender ? 'Yes' : 'No'}`);
        console.log();
      });
      
      return videos;
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Download Facebook video in best quality (HD preferred)
 * @param {string} url - Facebook video URL
 */
async function downloadFacebookVideoHD(url) {
  try {
    console.log('\n=== Downloading Facebook Video (Best Quality) ===\n');
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/facebook/download-hd', {
      params: { url }
    });
    
    const { success, video, allQualities } = response.data;
    
    if (success) {
      console.log('✅ Best Quality Download Ready!\n');
      console.log('🎬 Selected Quality:');
      console.log(`   Resolution: ${video.resolution}`);
      console.log(`   URL: ${video.url}`);
      console.log(`   Thumbnail: ${video.thumbnail}`);
      console.log(`   Should Render: ${video.shouldRender ? 'Yes' : 'No'}`);
      console.log(`\n📊 Available Qualities: ${allQualities.length}`);
      
      allQualities.forEach((q, i) => {
        console.log(`   ${i + 1}. ${q.resolution}`);
      });
      
      return video;
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Download multiple Facebook videos
 * @param {Array<string>} urls - Array of Facebook video URLs
 */
async function downloadMultipleVideos(urls) {
  try {
    console.log('\n=== Downloading Multiple Facebook Videos ===\n');
    
    const results = [];
    
    for (const url of urls) {
      try {
        const response = await api.get('/facebook/download-hd', {
          params: { url }
        });
        
        results.push({
          url,
          success: true,
          video: response.data.video
        });
        
        console.log(`✅ Downloaded: ${response.data.video.resolution}`);
      } catch (error) {
        results.push({
          url,
          success: false,
          error: error.response?.data?.error || error.message
        });
        
        console.log(`❌ Failed: ${url}`);
      }
    }
    
    console.log(`\n📋 Downloaded ${results.filter(r => r.success).length}/${urls.length} videos`);
    return results;
  } catch (error) {
    console.error('Error:', error.message);
    throw error;
  }
}

/**
 * Get specific quality from Facebook video
 * @param {string} url - Facebook video URL
 * @param {string} preferredResolution - Preferred resolution (e.g., "720p", "1080p")
 */
async function downloadSpecificQuality(url, preferredResolution) {
  try {
    console.log(`\n=== Downloading Facebook Video (${preferredResolution}) ===\n`);
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/facebook/download', {
      params: { url }
    });
    
    const { success, videos } = response.data;
    
    if (success) {
      // Find preferred quality
      const video = videos.find(v => 
        v.resolution.toLowerCase().includes(preferredResolution.toLowerCase())
      ) || videos[0];
      
      console.log('✅ Download URL Ready!\n');
      console.log(`Resolution: ${video.resolution}`);
      console.log(`URL: ${video.url}`);
      console.log(`Thumbnail: ${video.thumbnail}`);
      
      return video;
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Get video information without downloading
 * @param {string} url - Facebook video URL
 */
async function getVideoInfo(url) {
  try {
    console.log('\n=== Getting Facebook Video Info ===\n');
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/facebook/download', {
      params: { url }
    });
    
    const { success, videos } = response.data;
    
    if (success) {
      console.log('📊 Video Information:\n');
      console.log(`Available Qualities: ${videos.length}`);
      
      videos.forEach((video, index) => {
        console.log(`\n${index + 1}. ${video.resolution}`);
        console.log(`   Has Thumbnail: ${video.thumbnail ? 'Yes' : 'No'}`);
        console.log(`   Needs Rendering: ${video.shouldRender ? 'Yes' : 'No'}`);
      });
      
      // Find best quality
      const bestQuality = videos.find(v => v.resolution.includes('1080p')) ||
                         videos.find(v => v.resolution.includes('720p')) ||
                         videos[0];
      
      console.log(`\n🏆 Best Quality: ${bestQuality.resolution}`);
      
      return {
        totalQualities: videos.length,
        qualities: videos.map(v => v.resolution),
        bestQuality: bestQuality.resolution
      };
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Download and compare qualities
 * @param {string} url - Facebook video URL
 */
async function compareQualities(url) {
  try {
    console.log('\n=== Comparing Video Qualities ===\n');
    console.log(`URL: ${url}\n`);
    
    const response = await api.get('/facebook/download', {
      params: { url }
    });
    
    const { success, videos } = response.data;
    
    if (success) {
      console.log('📊 Quality Comparison:\n');
      
      const comparison = videos.map(video => ({
        resolution: video.resolution,
        hasThumbnail: !!video.thumbnail,
        needsRender: video.shouldRender,
        isDirectLink: !video.shouldRender
      }));
      
      console.table(comparison);
      
      return comparison;
    }
  } catch (error) {
    console.error('Error:', error.response?.data || error.message);
    throw error;
  }
}

// ===================
// EXAMPLES
// ===================

async function main() {
  try {
    const testUrl = 'https://www.facebook.com/share/r/14RStacRcih/';
    
    // Example 1: Download all quality options
    await downloadFacebookVideo(testUrl);
    
    // Example 2: Download best quality (HD preferred)
    await downloadFacebookVideoHD(testUrl);
    
    // Example 3: Download specific quality
    await downloadSpecificQuality(testUrl, '720p');
    
    // Example 4: Get video information
    await getVideoInfo(testUrl);
    
    // Example 5: Compare qualities
    await compareQualities(testUrl);
    
    // Example 6: Download multiple videos
    const urls = [
      'https://www.facebook.com/share/r/14RStacRcih/',
      'https://fb.watch/example123/'
    ];
    await downloadMultipleVideos(urls);
    
  } catch (error) {
    console.error('Main error:', error.message);
    process.exit(1);
  }
}

// Run examples
if (require.main === module) {
  main();
}

module.exports = {
  downloadFacebookVideo,
  downloadFacebookVideoHD,
  downloadMultipleVideos,
  downloadSpecificQuality,
  getVideoInfo,
  compareQualities
};
